### DISCUSSION TITLE
Network sandboxing...

### DISCUSSION THREAD

**Discussion by Tell-Sackett** [ruanformigoni/gameimage#95] [Ideas]
Would you consider making an easily toggleable option for preventing a flatimage from connecting online/to the internet?

As came out during the whole Unity price debacle: it has been shown that some companies behind these game engines are willing to bake (or at least consider doing so) telemetry into games made via their products. Additionally, as I was browsing GOG, I heard about a number of games (namely those published by PlayStation) with privacy policies which included mentions of telemetry data. One might argue that such practices raise serious privacy concerns.

While, from my understanding, it is possible to use an installed firewall on one's system for preventing such applications as flatimages from connecting to the internet, I can't help but think that a toggleable button with core access to the flatimage container might be much more efficient; such would also make the process much easier and straightforward (especially for less technical users).

**Comment by ruanformigoni** [comment]
Hi @Tell-Sackett , I am re-implementing flatimage in C++ and optional network access is already done :smile:. As soon as I update GameImage to use the novel version, I will create an option to toggle networking in the wizard.

**Reply by Tell-Sackett** [reply]
Sick; thanks! :thumbsup:

I remember back when the project was just in cli you had mentioned transitioning it to C++. After the C++ branch swapped to main, however, it seemed to start shifting over some to Rust. If you don't mind me asking: were you, at one point, thinking about switching it over to Rust but have now decided against such, or have you never stopped the C++ transition and just played with Rust on the side? Just curious; I'm currently attempting to learn C, and I'm trying to scope the landscape with Rust adoption flourishing like it is.

**Reply by ruanformigoni** [reply]
> never stopped the C++ transition and just played with Rust on the side

No problem, it was more like this. The main application was intended to be and is written is C++, I had some experience with rust previously and wanted to expand my knowledge of the language. GameImage is two separated applications, the main program (backend) is written in C++ and the GUI (frontend) in rust; the GUI communicates with the backend through a linux mechanism called [message queues](https://www.man7.org/linux/man-pages/man7/mq_overview.7.html).

In rust, I found several limitations that I do not like when compared to C++, things that would be very concise in C++ are very expressive in rust. For example:

Rust does not have an auto clone syntax yet, that makes everything very repetitive and verbose, there are macros to simplify, but it is only a workaround and auto completion stops working when that is used:

```rust
  let mut clone_output_status = ret_frame_footer.output_status.clone();
  let clone_label = label.to_string();
  let clone_frame_list = frame_list.clone();
  let clone_tx = tx.clone();
  btn_del.set_callback(move |_|
```

In c++ this could be a single line, assuming the objects are encapsulated in shared pointers.
```cpp
  btn_del.set_callback([=](...){ ... })
```

Rust plays it safe and becomes verbose because of it, it is still a great language, but new and with ample room for improvement. The problem with C++ that I think is the reason rust has become so successful is twofold: First, it is possible to write terrible code in C++ that mixes C and is completely unsafe and broken, it takes time to learn it properly and modern C++ with its amazing standard library is not what I usually see people teaching in online courses. Second, rust has a package manager shipped with it; C++ has them too (I use conan for gameimage), but they are not as simple to use, there is no "official" one so they are fragmented in adoption.

**Reply by Tell-Sackett** [reply]
Gotcha', cool stuff!

I didn't know about package managers for either of these languages. That sounds like it would make the process of learning each much easier (assuming they work properly and well). I know and can implement a number of C standard library functions, but my lack of knowledge on how to utilize external libraries (for graphical implementation) is really hampering my ability to make code which is user-accessible to those scared of the command line. I've been struggling with stuff even as simple as Raylib.

But yeah, I'm definitely not on that level right now. Gonna' need a lot more practice before I could actually hope to contribute to a project like this. That said, I think I might take a break from C for the time being and practice some GDScript... figured that could be a fun side-hobby. Hopefully I might still prove useful to the project as a guinea pig, if nothing else.
